### PR TITLE
Add utils to translate custom types

### LINKED ISSUE #826: Create utils to make it easier to work with custom types
**What would you like to be added**:
Helper funcs that help with type conversion, especially from and to pointers, for our custom types. Examples of these types include `Hostname`, `Namespace`, and `Group`.

**Why this is needed**:
To simplify CRD validation, we've created a lot of custom types. Unfortunately this makes it difficult to work with our APIs with go. It would be very helpful if we had some utils that made this easier to work with. Otherwise each implementation will likely end up implementing their own set of these.

### PR DESCRIPTION
**Pull request by abhijit-dev82**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Proposed change set are helper functions to translate to and from custom types.

**Which issue(s) this PR fixes**:
Fixes #826

**Does this PR introduce a user-facing change?**:
No

### CODE REVIEW

File: apis/v1beta1/helper.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (robscott) [inline apis/v1beta1/helper.go:2]:
```suggestion
Copyright 2022 The Kubernetes Authors.
```

File: apis/v1alpha2/helper.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (robscott) [inline apis/v1alpha2/helper.go:2]:
```suggestion
Copyright 2022 The Kubernetes Authors.
```

File: apis/v1beta1/helper.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	pn := PortNumber(port)
+	return &pn
+}
+
+func PortNumFromPtr(name *PortNumber) int32 {
+	portNum := int32(*name)
+	return portNum
+}
+
+func ListenerHostnameToPtr(host string) *Hostname {
```
Comment (robscott) [inline apis/v1beta1/helper.go:39]: Maybe remove `Listener` prefix here since Hostname can be used elsewhere.

File: apis/v1beta1/helper.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+func SectionNameToPtr(sectionName string) *SectionName {
```
Comment (robscott) [inline apis/v1beta1/helper.go:19]: I think most go linting will usually fail without comments describing each exported function

**Review by robscott — COMMENTED** [review]
Thanks @abhijit-dev82! This looks like a great start. Will discuss this at the community meeting on Monday to ensure that at least some implementations will actually make use of this. This seems like it could be very helpful, but don't want to introduce extra code to maintain until we're sure it will be used.

**Comment by carlisia** [comment]
@abhijit-dev82 thank you for doing this, I will certainly use it so I can delete some code on my code base!

Everything lgtm except I'm not loving the name and the location of the file, especially when I consider existing helper functionality.

I opened an issue to explain: [Consider a minor restructure of directory/file for util things · Issue #1336 · kubernetes-sigs/gateway-api](https://github.com/kubernetes-sigs/gateway-api/issues/1336). I'd hate to block the merging of this PR for something that is sorta trivial but brakes the interface so it will require some thinking. At the same time, I would love adding such helpful code in a place that makes great sense, given the context around it.

**Comment by robscott** [comment]
I really like @carlisia's idea in #1336. @abhijit-dev82 do you think you could incorporate that restructuring into this PR? 

/ok-to-test

**Comment by carlisia** [comment]
There is one more input I'd like to offer. If we look at the existing conversion helpers (`PathMatchTypePtr` and `PortNumberPtr` here:

https://github.com/kubernetes-sigs/gateway-api/blob/854e2bfc52764bda4d81dd24851ab124bf564e8f/apis/v1alpha2/validation/util/utils.go#L24  

...we see that there is no `To` or `From`. That style is idiomatic of Go because it doesn't duplicate information that we can see in a different way. In this case, that `PathMatchTypePtr` takes a string is indicated in the signature, and it is idiomatic to name the function based on what it returns, in this case the pointer. Sooooo.... long paragraph to suggest changing the names to fit this format:

`s/SectionNameToPtr/SectionNamePtr`
`s/SectionNameFromPtr/SectionNameStr`

File: apis/v1alpha2/helper.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	pn := PortNumber(port)
+	return &pn
+}
+
+func PortNumFromPtr(name *PortNumber) int32 {
+	portNum := int32(*name)
+	return portNum
+}
+
+func ListenerHostnameToPtr(host string) *Hostname {
```
[the remark this answers is not in the document]
Comment (abhijit-dev82) [reply apis/v1alpha2/helper.go:39]: corrected

File: apis/v1beta1/helper.go (hunk 1 above)
Comment (abhijit-dev82) [reply apis/v1beta1/helper.go:2]: corrected.

File: apis/v1alpha2/helper.go (hunk 2 above)
Comment (abhijit-dev82) [reply apis/v1alpha2/helper.go:2]: Corrected.

**Comment by carlisia** [comment]
This PR adds the directory structure mentioned above: https://github.com/kubernetes-sigs/gateway-api/pull/1337?w=1.

**Comment by abhijit-dev82** [comment]
@carlisia & @robscott Thank you for all the comments. I was away due to health reasons, hence was not able to make the necessary changes. I see that the directory changes are already done, will add the new file in the util dir. Do we need to see we can auto generate these translation routines based on the custom types defined .

**Comment by abhijit-dev82** [comment]
HI @carlisia & @robscott , Had to open a new PR as was facing merge issues with the older branch. Below is the link
https://github.com/kubernetes-sigs/gateway-api/pull/1352

**Pull request closed without merging**